pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1323 @@
+					if T::AMM::exists(asset_pair) {
+						let amm_token = T::AMM::get_share_token(asset_pair);
+
+						let pallet_account = Self::account_id();
+						T::MultiCurrency::transfer(amm_token, &pallet_account, &who, deposit.shares)?;
+
+						//NOTE: Theoretically neither `GlobalPool` nor
+						//`LiquidityPoolYieldFarm` may not exist
+						Self::deposit_event(Event::SharesWithdrawn(who.clone(), amm_token, deposit.shares));
+					}
+
+					*maybe_nft = None;
+					pallet_nft::Pallet::<T>::do_burn(who, nft_class_id, nft_id)?;
+
+					if nfts_in_class.is_one() && can_destroy_nft_class {
+						*maybe_nft_class = None;
+						pallet_nft::Pallet::<T>::do_destroy_class(nft_class_id)?;
+					} else {
+						*maybe_nft_class = Some((
+							asset_pair,
+							nfts_in_class.checked_sub(1).ok_or(Error::<T>::Overflow)?,
+							farm_id,
+						));
+					}
+					Ok(().into())
+				})
+			})
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn get_next_id() -> Result<PoolId, Error<T>> {
+		PoolIdSeq::<T>::try_mutate(|current_id| {
+			*current_id = current_id.checked_add(1).ok_or(Error::<T>::Overflow)?;
+
+			Ok(*current_id)
+		})
+	}
+
+	/// Account id of pot holding all the shares
+	fn account_id() -> AccountIdOf<T> {
+		T::PalletId::get().into_account()
+	}
+
+	/// Return pallet account or pool acocunt from PoolId
+	///
+	/// WARN: pool_id = 0 is same as `T::PalletId::get().into_account()`. 0 is not valid value
+	fn pool_account_id(pool_id: PoolId) -> Result<AccountIdOf<T>, Error<T>> {
+		Self::validate_pool_id(pool_id)?;
+
+		Ok(T::PalletId::get().into_sub_account(pool_id))
+	}
+
+	/// Return now period number
+	fn get_now_period(blocks_per_period: BlockNumberFor<T>) -> Result<PeriodOf<T>, Error<T>> {
+		Self::get_period_number(T::BlockNumberProvider::current_block_number(), blocks_per_period)
+	}
+
+	/// Return period number from block number now and number of blocks in one period
+	fn get_period_number(
+		block: BlockNumberFor<T>,
+		blocks_per_period: BlockNumberFor<T>,
+	) -> Result<PeriodOf<T>, Error<T>> {
+		block.checked_div(&blocks_per_period).ok_or(Error::<T>::Overflow)
+	}
+
+	/// Loyalty multiplier  
+	///
+	// theta = periods/[(initial_reward_percentage + 1) * scale_coef];
+	//
+	// loyalty_multiplier = [theta + (theta * initial_reward_percentage) + initial_reward_percentage]/[theta + (theta * initial_reward_percentage) + 1]
+	//
+	fn get_loyalty_multiplier(periods: PeriodOf<T>, curve: Option<LoyaltyCurve>) -> Result<FixedU128, Error<T>> {
+		let curve = match curve {
+			Some(v) => v,
+			None => return Ok(FixedU128::one()),
+		};
+
+		//b.is_one() is special case - this case is prevented by loyalty curve params validation
+		if FixedPointNumber::is_one(&curve.initial_reward_percentage) {
+			return Ok(FixedU128::one());
+		}
+
+		let denom = curve
+			.initial_reward_percentage
+			.checked_add(&1.into())
+			.ok_or(Error::<T>::Overflow)?
+			.checked_mul(&FixedU128::from(Into::<u128>::into(curve.scale_coef)))
```

Comment:
nitpick.this should work too.

```suggestion
			.checked_mul(&FixedU128::from(curve.scale_coef as u128))
```